### DISCUSSION TITLE
Set Product and Folding in C++

### DISCUSSION THREAD

**Discussion by soheil647** [cvc5/cvc5#10260] [Q&A] [answered]
I am struggling with creating a trick to have a set product with cpp API. 

Imagine I have 2 sets, SetA and SetB.
I can use a nested Set_MAP to create a Set of Sets where each inner set combines each element of SetA with all members of SetB. It would results in many Set_Unions which creates the Set of Sets. Now the problem is in the folding this Set of Sets into a single Set. Using CVC5 set folding It requires a lambda function with 2 argument that the second argument should be the accumulator set. However I could not find a way to declare such LAMBDA function as it only takes 1 argument.

Bags product is implemented in CVC5, however, converting sets to Bag and converting back the Bags to Sets is not supported when asserting any formula containing such conversions. Thus, this option is not possible.

I would greatly appreciate any help with implementing a SET Cartesian product such as one implemented with Relations or Bags or any idea that helps me in this way. My goal is to use cvc5 for set of (unique) tuples. 

For nested maps I have:

    std::vector<Term> newTupleElements;
    DatatypeConstructor ctorA = setA.getSort().getSetElementSort().getDatatype()[0];
    for (int i = 0; i < 5; ++i) {
        const DatatypeSelector& selector = ctorA[i];
        newTupleElements.push_back(slv.mkTerm(Kind::APPLY_SELECTOR, {selector.getTerm(), boundVarA}));
    }
    DatatypeConstructor ctorB = setB.getSort().getSetElementSort().getDatatype()[0];
    for (int i = 0; i < 3; ++i) {
        const DatatypeSelector& selector = ctorB[i];
        newTupleElements.push_back(slv.mkTerm(Kind::APPLY_SELECTOR, {selector.getTerm(), boundVarB}));
    }
    cout << newTupleElements << endl;

    Term concatenatedTuple = slv.mkTuple(newTupleElements);

    Term lambdaInner = slv.mkTerm(Kind::LAMBDA, {varListB, concatenatedTuple});
    // Apply the first map
    Term firstMap = slv.mkTerm(Kind::SET_MAP, {lambdaInner, setB});
    cout << firstMap << endl;
    // Define a lambda for the outer map
    // This lambda applies the first map to each element of setA
    Term lambdaOuter = slv.mkTerm(Kind::LAMBDA, {varListA, firstMap});
    cout << lambdaOuter << endl;
    // Apply the outer map to create the Cartesian product
    Term cartesianProduct = slv.mkTerm(Kind::SET_MAP, {lambdaOuter, setA});
    
However, the folding is not working correctly:

    Sort setOfSetsSort = cartesianProduct.getSort();
    Sort innerSetSort = setOfSetsSort.getSetElementSort();
    Term boundVarAB = slv.mkVar(innerSetSort, "rowAB");
    std::vector<Term> varsAB = {boundVarAB};
    Term varListAB = slv.mkTerm(Kind::VARIABLE_LIST, varsAB);

    Term emptySet = slv.mkEmptySet(cartesianProduct.getSort());

    // Create a lambda function for the folding operation
    Term lambdaVarSet = slv.mkVar(innerSetSort, "innerSet"); // Variable for the inner set
    Term lambdaVarAccum = slv.mkVar(innerSetSort, "accumSet"); // Variable for the accumulating set
    cout << lambdaVarSet.getSort() << endl;

    // Define the union operation as the folding function
    Term unionExpr = slv.mkTerm(Kind::SET_UNION, {boundVarAB, boundVarAB});
    Term lambdaUnion = slv.mkTerm(Kind::LAMBDA, {varListAB, unionExpr});
    cout << endl << endl;
    cout << lambdaUnion << endl;

    // Apply the SET_FOLD operation
    Term flattenedSet = slv.mkTerm(Kind::SET_FOLD, {lambdaUnion, emptySet, cartesianProduct});

    std::cout << "Flattened Set: \n" << flattenedSet << std::endl;

Getting error of:

 > libc++abi: terminating with uncaught exception of type cvc5::CVC5ApiException: Operator set.fold expects a function of type  (-> (Set (Tuple Int String String String String String String String)) T2 T2). Found a function of type '(-> (Set (Tuple Int String String String String String String String)) (Set (Tuple Int String String String String String String String)))'.

which make sense since LAMBDA function only takes 2 inputs not 3. Any help with this matter is appreciated.

Thanks a lot in advance.

**Comment by mudathirmahgoub** [comment]
Did you try `Kind::RELATION_PRODUCT` ?

**Comment by mudathirmahgoub** [comment, marked as answer]
Answered in this issue
https://github.com/cvc5/cvc5/issues/10261
